build: manage thirdparty & zetasql via cmake

aceforeverd opened the pull request:
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

introduce new way to manage thirdparty dependencies.
it add a cmake module `third-party` to manage dependencies which resolve following issues:

- #550 #594 
- #636 : run maven with `--batch-mode` option in CI runner
- #625 : now we can compile openmldb once, no more hybridse install -> openmldb



* **What is the current behavior?** (You can also link to an open issue here)

output if runs `cmake -S . -B build`, `build/usr/` is the fiinal install dir
```
third-party/
├── CMakeLists.txt
└── build
    ├── CMakeCache.txt
    ├── CMakeFiles
    ├── Makefile
    ├── build
    │   ├── downloads
    │   ├── src
    │   └── tmp
    ├── cmake_install.cmake
    └── usr
        ├── bin
        ├── certs
        ├── include
        ├── lib
        ├── misc
        ├── openssl.cnf
        ├── openssl.cnf.dist
        ├── private
        └── share
```


* **What is the new behavior (if this is a feature change)?**



* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

old dependency setup script deprecated, instead use `make thirdparty`

## Further work

- rm old dependency script e.g `init_env.sh`
- toydb in hybridse
- build zetasql source from commit has, it avoid rebuilding every time
- support compile from source WIP: https://github.com/aceforeverd/fedb/tree/test/618-build-thirdparty

aceforeverd committed:
build: create cmake sub directory in third-party

aceforeverd committed:
build(hybridse): adopt third-party cmake module

aceforeverd committed:
build: add thirdparty.sh

aceforeverd committed:
build(hybridse): update scripts

aceforeverd committed:
build: fix openmldb build

aceforeverd committed:
build(third-party): add hybridsql asserts

aceforeverd committed:
revert: scripts in steps

aceforeverd committed:
fix: init_env.sh

aceforeverd committed:
fix: init_env.sh

aceforeverd commented on hybridse/CMakeLists.txt:
```diff
@@ -52,8 +52,36 @@ option(JAVASDK_ENABLE "Enable javasdk" ON)
+if(DEFINED ENV{DEPS_BUILD_DIR})
+    set(DEPS_PREFIX "$ENV{DEPS_BUILD_DIR}/usr" CACHE PATH "Path prefix for finding dependencies")
+else()
+    set(DEPS_PREFIX "${CMAKE_CURRENT_SOURCE_DIR}/.deps/usr" CACHE PATH "Path prefix for finding dependencies")
```
add .deps to ignorelist

aceforeverd commented on third-party/CMakeLists.txt:
```diff
@@ -0,0 +1,95 @@
+  message(STATUS "Download pre-compiled zetasql from ${ZETASQL_URL}")
+  # download pre-compiled zetasql from GitHub Release
+  ExternalProject_Add(zetasql
+    URL ${ZETASQL_URL}
```
ULR_HASH

aceforeverd commented on hybridse/CMakeLists.txt:
```diff
@@ -52,8 +52,36 @@ option(JAVASDK_ENABLE "Enable javasdk" ON)
 option(EXAMPLES_ENABLE "Enable examples" ON)
 option(LLVM_EXT_ENABLE "Enable llvm ext sources" OFF)
 
+# thirdparty snippets inspired from neovim: https://github.com/neovim/neovim/blob/master/CMakeLists.txt
```
rm this

aceforeverd commented on third-party/CMakeLists.txt:
```diff
@@ -0,0 +1,95 @@
+    BUILD_COMMAND ""
+    INSTALL_COMMAND "")
+  ExternalProject_Get_Property(hybridsql-asserts DOWNLOADED_FILE)
+  add_custom_target(hybridsql-asserts-extract ALL tar xzf ${DOWNLOADED_FILE} -C ${DEPS_INSTALL_DIR} --strip-components=1
```
avoid unpack if same version

aceforeverd replied:
#614

aceforeverd commented on third-party/CMakeLists.txt:
```diff
@@ -0,0 +1,95 @@
+    COMMENT "Extracting zetasql ${DOWNLOADED_FILE} into ${DEPS_INSTALL_DIR}")
+  add_dependencies(zetasql-extract zetasql)
+else()
+  message(STATUS "Compile zetasql from source: ${ZETASQL_HOME}@${ZETASQL_TAG}")
```
find_program(bazel)

aceforeverd commented on third-party/CMakeLists.txt:
```diff
@@ -0,0 +1,95 @@
+
+include(ExternalProject)
+
+set(DEPS_INSTALL_DIR "${CMAKE_BINARY_DIR}/usr" CACHE PATH "Dependencies install directory.")
```
set this option

aceforeverd committed:
build(cmake): update third-party module

aceforeverd committed:
build: add global makefile

aceforeverd committed:
build: update hybridse ci

aceforeverd committed:
ci: update cicd.yaml

aceforeverd committed:
ci: fix build

aceforeverd committed:
Merge remote-tracking branch 'upstream/main' into build/add-zetasql-to-cmake

aceforeverd committed:
fix: Makefile

aceforeverd committed:
ci: fix cicd.yaml

aceforeverd committed:
build: resolve #614

zhanghaohit commented on third-party/cmake/FetchZetasql.cmake:
```diff
@@ -0,0 +1,60 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set(ZETASQL_HOME https://github.com/jingchen2222/zetasql)
```
Can move this repo to [4paradigm](https://github.com/4paradigm/)?

aceforeverd replied:
@jingchen2222 it should be. For some historic reason we don't transfer the forked repo to 4Paradigm

jingchen2222 replied:
Can update zetasql_home now?

aceforeverd replied:
yes

aceforeverd committed:
ci: fix cicd.yaml

delete thirdparty build directory to save space

aceforeverd committed:
chore: update comments

aceforeverd committed:
build: fix Makefile

aceforeverd committed:
docs: add third-party readme

aceforeverd committed:
docs: update compile document

aceforeverd committed:
Merge remote-tracking branch 'upstream/main' into build/add-zetasql-to-cmake

aceforeverd committed:
chore: add a TODO note in Makefile

aceforeverd committed:
Merge remote-tracking branch 'upstream/main' into build/add-zetasql-to-cmake

aceforeverd committed:
Merge remote-tracking branch 'upstream/main' into build/add-zetasql-to-cmake

aceforeverd committed:
Merge remote-tracking branch 'upstream/main' into build/add-zetasql-to-cmake

aceforeverd committed:
chore: update gitignore

aceforeverd committed:
Merge remote-tracking branch 'upstream/main' into build/add-zetasql-to-cmake

aceforeverd committed:
wow

aceforeverd committed:
Merge remote-tracking branch 'upstream/main' into build/add-zetasql-to-cmake

aceforeverd committed:
build: update compile steps

- more information building third-party
- run maven in batch mode when on CI environment
- configure javadoc plugin: set 'quiet' to true

aceforeverd committed:
build: tweak some maven plugin options

- javadoc plugin: silent = true
- git commit plugin: verbose = false

aceforeverd committed:
build: set default build type to RelWithDebInfo

workaround to #748

aceforeverd committed:
fix compile

aceforeverd committed:
ci: fix environment variable

aceforeverd committed:
feat: resolving #625

add hybridse to root CMakeLists.txt

aceforeverd committed:
fix(makefile): configure depends on thirdparyt

aceforeverd committed:
build: set MAVEN_FLAGS in hybridse as well

aceforeverd committed:
docs: update compile and third-party README

aceforeverd committed:
Merge remote-tracking branch 'upstream/main' into build/add-zetasql-to-cmake

aceforeverd commented on hybridse/tools/compile_and_coverage.sh:
```diff
@@ -24,9 +24,7 @@ popd
-# install thirdparty hybrise
-HYBRIDSE_THIRDPARTY="$(pwd)/thirdparty"
-../steps/setup_thirdparty.sh "${HYBRIDSE_THIRDPARTY}"
+./tools/third-party.sh
```
deprecating those scripts

aceforeverd commented on java/pom.xml:
```diff
@@ -608,6 +614,20 @@
+    <repositories>
+      <repository>
+        <id>maven-snapshots</id>
+        <url>https://s01.oss.sonatype.org/content/repositories/snapshots</url>
```
snapshot repository may not needed

aceforeverd committed:
build: cleanup scripts

aceforeverd committed:
fix: java coverage

zhanghaohit commented on steps/init_env.sh:
```diff
@@ -50,15 +50,15 @@ echo "HYBRIDSE_SOURCE: $HYBRIDSE_SOURCE"
 if [[ ${HYBRIDSE_SOURCE} = "local" ]]; then
```
steps/init_env.sh and steps/setup_thirdparty.sh are still needed?

aceforeverd replied:
it may still required from https://github.com/4paradigm/OpenMLDB/blob/main/.github/workflows/integration-test-src.yml. 
Mark as further work

aceforeverd commented on hybridse/examples/toydb/src/CMakeLists.txt:
```diff
@@ -59,6 +59,7 @@ if (TESTING_ENABLE AND EXAMPLES_TESTING_ENABLE)
         file(RELATIVE_PATH RELATIVE_TEST_PATH ${TEST_CPP_SOURCE_DIR} ${TEST_SCRIPT})
         get_filename_component(TEST_TARGET_DIR ${RELATIVE_TEST_PATH} DIRECTORY)
         get_filename_component(TEST_TARGET_NAME ${RELATIVE_TEST_PATH} NAME_WE)
+        set(TEST_TARGET_NAME "toydb_${TEST_TARGET_NAME}")
```
Further work: keep toydb ?

aceforeverd commented on third-party/cmake/FetchZetasql.cmake:
```diff
@@ -0,0 +1,63 @@
+if (CMAKE_SYSTEM_NAME STREQUAL "Linux")
+  if (CMAKE_SYSTEM_PROCESSOR MATCHES "(arm64)|(ARM64)|(aarch64)|(AARCH64)")
+    # aarch64 support is experimental
+    set(ZETASQL_URL "https://github.com/aceforeverd/zetasql/releases/download/v0.2.1.beta9/libzetasql-0.2.1.beta9-linux-gnu-aarch64.tar.gz")
```
rm arm64 now

aceforeverd committed:
build: rm zetasql precompiled for arm64

tobegit3hub reviewed (approved):
LGTM
